### PR TITLE
add compression method for files: Xz

### LINKED ISSUE #8828: Generic compression methods for files: add `xz`.
**Use case**
Directly import test data from
https://clickhouse-datasets.s3.yandex.net/hits/tsv/hits_v1.tsv.xz
(https://clickhouse.yandex/docs/en/getting_started/example_datasets/metrica/)
inside `clickhouse-client`.

**Describe the solution you'd like**
Similar to `gz` and `br`.

**Additional context**
`xz` is extremely slow in compression and moderate slow in decompression. Decompression speed almost independent on compression level. `xz` is a good alternative to bzip. We can add `xz` and avoid support for `bz2`.

This is not intended for data storage in tables. Only for import/export and backups.

### PR DESCRIPTION
**Pull request by fibersel**
Changelog category (leave one):

- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
add `*.xz` compression/decompression support.It enables using `*.xz` in `file()` function.This closes [#8828](https://github.com/ClickHouse/ClickHouse/issues/8828)

### CODE REVIEW

File: contrib/xz-cmake/CMakeLists.txt
Comment (azat) [inline contrib/xz-cmake/CMakeLists.txt:246]: Does all of this should be public includes? (i.e. the one that will be added while building library itself and those who linked against it)

File: contrib/xz-cmake/CMakeLists.txt
Comment (azat) [inline contrib/xz-cmake/CMakeLists.txt:234]: But there is no users of this library, I guess that is the reason for an error:

```
/home/centos/ClickHouse/src/IO/LzmaReadBuffer.h:7:10: fatal error: lzma.h: No such file or directory
7 | #include <lzma.h>
```

Comment (azat) [reply contrib/xz-cmake/CMakeLists.txt:234]: I guess it should be added in the same manner as here - https://github.com/ClickHouse/ClickHouse/blob/90dcc135b4043ecca726adc2dc9470ef48a0fba2/src/CMakeLists.txt#L325-L330

File: contrib/xz-cmake/CMakeLists.txt
Comment (azat) [inline contrib/xz-cmake/CMakeLists.txt:1]: But `xz` library already contains CMakeLists is it unusable?

**Comment by azat** [comment]
>Where can I find tests for ZlibInflatingReadBuffer.h?I grepped tests/ folder and found nothing

```
src/IO/tests/zlib_buffers.cpp:#include <IO/ZlibInflatingReadBuffer.h>
src/IO/tests/zlib_buffers.cpp:        DB::ZlibInflatingReadBuffer inflating_buf(std::move(buf), DB::CompressionMethod::Gzip);
```

File: contrib/xz-cmake/CMakeLists.txt
Comment (fibersel) [reply contrib/xz-cmake/CMakeLists.txt]: reused `xz` CMakeLists

File: contrib/xz-cmake/CMakeLists.txt
Comment (fibersel) [reply contrib/xz-cmake/CMakeLists.txt]: Did it in similar fashion

File: contrib/xz-cmake/CMakeLists.txt
Comment (fibersel) [reply contrib/xz-cmake/CMakeLists.txt]: Completely removed extra folder for CMake, exposed for includes only one folder `liblzma/api`

File: src/IO/LzmaWriteBuffer.cpp
Comment (azat) [inline src/IO/LzmaWriteBuffer.cpp:109]: Looks like indent is wrong

File: src/IO/LzmaWriteBuffer.cpp
Comment (azat) [inline src/IO/LzmaWriteBuffer.cpp:87]: Forgot to remove? (and in a few other places)

File: src/IO/LzmaWriteBuffer.cpp
Comment (azat) [inline src/IO/LzmaWriteBuffer.cpp:80]: Coding style: curly bracket is always on a new line (and not only here but in a few other places) (except for `} while (true)`)

File: src/IO/LzmaReadBuffer.cpp
Comment (azat) [inline src/IO/LzmaReadBuffer.cpp:69]: This can use fmt-like constructor

```suggestion
            throw Exception(ErrorCodes::LZMA_STREAM_DECODER_FAILED, "lzma decoder finished, but stream is still alive: error code: {}; lzma version: {}", ret, LZMA_VERSION_STRING);
```

**Review by azat — COMMENTED** [review]
Just a few minor comments until there is no can-be-tested label

File: src/IO/LzmaWriteBuffer.cpp
Comment (azat) [inline src/IO/LzmaWriteBuffer.cpp:105]: trailing whitespace, and incorrect indent below (tabs instead of spaces, and tabs also not only here)

And by the way it is not only here, it is pretty easy to fix just using git, but please note that this will override the history (and that said that if you don't know git I guess it will be better if you will fix this in your editor)

```
git rebase --whitespace=fix upstream/master
```

File: contrib/CMakeLists.txt
Comment (azat) [inline contrib/CMakeLists.txt:34]: Looks like you've forget to add submodule

Comment (fibersel) [reply contrib/CMakeLists.txt:34]: Isn't it?
https://github.com/ClickHouse/ClickHouse/pull/16578/files#diff-fe7afb5c9c916e521401d3fcfb4277d5071798c3baf83baf11d6071742823584R193

Comment (azat) [reply contrib/CMakeLists.txt:34]: Not really, this is just to know where to fetch the submodule from.

You need to run smth like this:

```
git submodule add https://github.com/xz-mirror/xz contrib/xz
```

And this will create a special file `contrib/xz` in the git index (but in the filesystem you will see it as a content of the repository), that will contain the HEAD for the submodule

File: src/IO/LzmaWriteBuffer.cpp
Comment (fibersel) [reply src/IO/LzmaWriteBuffer.cpp:105]: Can I just format it with `clang-format` and options from .clang-format?

Comment (azat) [reply src/IO/LzmaWriteBuffer.cpp:105]: I guess (I've never used it)

**Comment by fibersel** [comment]
I kind of fixed it with clang-format.I manually look for tabs with vim and don't find it.Looks like all whitespaces consist of single spaces now.

File: contrib/CMakeLists.txt
Comment (azat) [reply contrib/CMakeLists.txt:34]: Also to pass fasttest you need to add `contrib/xz` into https://github.com/ClickHouse/ClickHouse/blob/f10a5207f4e89f73779375a042185b64b0b8db39/docker/test/fasttest/run.sh#L130

File: src/CMakeLists.txt
Comment (azat) [inline src/CMakeLists.txt:333]: And I guess it is better to move this out into separate `contrib/xz-cmake/CMakeLists.txt`, and add an interface library.
One of similar files is `contrib/protobuf-cmake/CMakeLists.txt` (protobuf has it's own cmake rules, while in clickhouse there is just a wrapper that install some options)

Comment (fibersel) [reply src/CMakeLists.txt:333]: What should I do to pass description check?
I've added record to `CHANCHELOG.md`

Comment (azat) [reply src/CMakeLists.txt:333]: >What should I do to pass description check?

Update description of PR and include changelog entry using specified format.
There is template for PRs - https://raw.githubusercontent.com/ClickHouse/ClickHouse/master/.github/PULL_REQUEST_TEMPLATE.md

Just copy it and modify the template, but leave the format.

**Comment by fibersel** [comment]
Ok, it seems, that I should write a wrapper for CMake

**Comment by fibersel** [comment]
It seems like `tuklib_cpucores` calls additional function that sets `TUKLIB_CPUCORES_DEFINITIONS` to `TUKLIB_CPUCORES_SCHED_GETAFFINITY`, but after that check still fails.
https://github.com/xz-mirror/xz/blob/master/cmake/tuklib_cpucores.cmake#L39

**Comment by azat** [comment]
>It seems like tuklib_cpucores calls additional function that sets TUKLIB_CPUCORES_DEFINITIONS to TUKLIB_CPUCORES_SCHED_GETAFFINITY, but after that check still fails.

The reason is that `if (DEFINED CACHED{VAR})` is added only in [3.14](https://cmake.org/cmake/help/v3.14/release/3.14.html)

You can try upgrade fasttest image to 20.04 (there are few images that already uses it anyway, so I guess it is fine, @alesapin ?) and it will contain cmake 3.16
https://github.com/ClickHouse/ClickHouse/blob/268c80520fb13fdeaaff6085d8705fb470f142cd/docker/test/fasttest/Dockerfile#L2

P.S. you can also send a patch to the uptream to bump cmake requirements:
https://github.com/xz-mirror/xz/blob/869b9d1b4edd6df07f819d360d306251f8147353/CMakeLists.txt#L50

**Comment by alexey-milovidov** [comment]
We usually write our own CMakeLists for libraries. Because original CMakeLists are intended to build a library itself, not to build it for another project.

**Comment by fibersel** [comment]
> `2020-11-04 01:46:00 -- Using CMake version 3.13.4`
Is seems like after updating base image to version 20.04 cmake is not being updated to 3.16

**Comment by azat** [comment]
>Is seems like after updating base image to version 20.04 cmake is not being updated to 3.16

Yeah, this is because fasttest image is not updated on pr-PR bases (as most of the images)

**Comment by fibersel** [comment]
What should I do further?
Fast test fails due to old image, style check is empty

**Comment by alexey-milovidov** [comment]
For Style check, see "output.txt".

**Comment by alexey-milovidov** [comment]
Please merge with master again...

File: CHANGELOG.md
Comment (alexey-milovidov) [inline CHANGELOG.md:222]: No, it's not for 20.9.
It will be automatically generated for the next release (20.12).

File: contrib/CMakeLists.txt
Comment (alexey-milovidov) [reply contrib/CMakeLists.txt:39]: Also it's possible to disable `xz` in fasttest.

File: src/Common/ErrorCodes.cpp
Comment (alexey-milovidov) [inline src/Common/ErrorCodes.cpp:527]: It should be near 552,
1000+ are for "other exceptions".

File: src/CMakeLists.txt
Comment (alexey-milovidov) [reply src/CMakeLists.txt:333]: Yes, it's much better to provide our own CMakeLists.

CMake often cannot be safely reused (without building unneeded targets, polluting build options).

File: src/IO/LzmaReadBuffer.cpp
Comment (alexey-milovidov) [inline src/IO/LzmaReadBuffer.cpp:1]: Just a small thing... in our style we keep capital letters in abbreviations, eg.
LZMA, not Lzma
JSON, not Json
HTTP, not Http.

File: src/IO/LzmaReadBuffer.cpp
Comment (alexey-milovidov) [inline src/IO/LzmaReadBuffer.cpp:19]: Motivation is unclear.

File: src/IO/LzmaReadBuffer.cpp
Comment (alexey-milovidov) [inline src/IO/LzmaReadBuffer.cpp:20]: If you write 500ULL it will allow to avoid errors if someone will change it to 5000
(because 500 is int, not uint64_t).

File: src/IO/LzmaReadBuffer.cpp
Comment (alexey-milovidov) [inline src/IO/LzmaReadBuffer.cpp:25]: There is fmt-style constuctor:

`Exception(ErrorCodes::LZMA_STREAM_DECODER_FAILED, "template {} {}", var1, var2)`

File: src/IO/LzmaReadBuffer.h
Comment (alexey-milovidov) [inline src/IO/LzmaReadBuffer.h:4]: Unused header.

File: src/IO/LzmaReadBuffer.h
Comment (alexey-milovidov) [inline src/IO/LzmaReadBuffer.h:31]: This field is not really needed.

File: src/IO/LzmaReadBuffer.h
Comment (alexey-milovidov) [inline src/IO/LzmaReadBuffer.h:15]: There is inconsistency between the names:

ZlibInflatingReadBuffer
ZlibDeflatingReadBuffer
LZMAReadBuffer
LZMAWriteBuffer

The inflating and deflating clarification makes the names more clear.
Because it's actually possible to make a WriteBuffer that will decompress data before writing instead of compressing.

File: src/IO/LzmaWriteBuffer.cpp
Comment (alexey-milovidov) [inline src/IO/LzmaWriteBuffer.cpp:112]: fmt-style is better.

File: src/IO/LzmaWriteBuffer.cpp
Comment (alexey-milovidov) [inline src/IO/LzmaWriteBuffer.cpp:80]: fmt-style is better.

File: src/IO/LzmaWriteBuffer.cpp
Comment (alexey-milovidov) [inline src/IO/LzmaWriteBuffer.cpp:27]: Two empty lines is too many for function body.

File: src/IO/LzmaWriteBuffer.cpp
Comment (alexey-milovidov) [inline src/IO/LzmaWriteBuffer.cpp:29]: Missing comments (explanations), the code is unclear to me.

**Review by alexey-milovidov — COMMENTED** [review]
We also need a functional test.

grep 'Content-Encoding'
in `test/queries/0_stateless`

We need to make sure that `xz` tool can read our xz-compressed data
and we can read data compressed with the `xz` tool.

File: src/IO/tests/lzma_buffers.cpp
Comment (alexey-milovidov) [inline src/IO/tests/lzma_buffers.cpp:12]: Did you run it? (And what are the results in comparison with `xz`?)

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Ok, only some small changes required.

**Comment by alexey-milovidov** [comment]
Maybe we can also add a test for `file` table function?

**Comment by alexey-milovidov** [comment]
You can find similar tests in the following way:
- look at the git blame for the code related to CompressionMethod;
- copy sha of commit;
- search the commit on GitHub;
- open "issues" - there are also pull requests;
- now you have found pull requests where the feature was added;
- there should be the related tests.

File: src/IO/tests/lzma_buffers.cpp
Comment (fibersel) [reply src/IO/tests/lzma_buffers.cpp:12]: Yes, I ran. But I suspect, that my implementation is not completely optimal:
`(base) [centos@clickhouse build]$ ./src/IO/tests/lzma_buffers
Writing done. Elapsed: 12.78 s., 6.17 MB/s
Reading done. Elapsed: 0.79 s., 100.24 MB/s`

`(base) [centos@clickhouse build]$ ./src/IO/tests/zlib_buffers
Writing done. Elapsed: 8.04 s., 110.55 MB/s
Reading done. Elapsed: 4.44 s., 200.08 MB/s`

Comment (alexey-milovidov) [reply src/IO/tests/lzma_buffers.cpp:12]: It's hard to say... Better to compare with `xz` tool, and if performance should not be significantly lower.

**Comment by fibersel** [comment]
`HTTPHandlers.cpp ` has too many changed because I did clang-format -i src/Server/HTTPHandler.cpp before commiting.

File: src/IO/CompressionMethod.cpp
Comment (alexey-milovidov) [inline src/IO/CompressionMethod.cpp:62]: A small note... `xz` is actually LZMA2. But why do we need this name? Maybe just leave `xz` only.

Comment (fibersel) [reply src/IO/CompressionMethod.cpp:62]: Is there the only place to remove LZMA?
Is it ok to leave all buffer filenames with LZMA* prefix?

**Review by alexey-milovidov — APPROVED** [review]
LGTM.

There are some leftover comments that are related to another xz implementation... we can try it in the next PR.

**Comment by fibersel** [comment]
> LGTM.
> 
> There are some leftover comments that are related to another xz implementation... we can try it in the next PR.

I had sent mail to contributor of fast-lzma2 maintainer, but he has not responded yet.I am not sure that fast-lzma2 is compatible with classic *.xz format and it's repository does not contain CMake file, so we need to adapt their implementation.I have faced some problems with that.Can we merge this implementation and keep in mind existence of another implementation?

**Comment by alexey-milovidov** [comment]
Yes, we can merge before trying fast-lzma2.

**Comment by alexey-milovidov** [comment]
You can remove commented code before we'll merge.

**Comment by alexey-milovidov** [comment]
I've tested decompression performance, it's the same as for `xz` tool.

**Comment by alexey-milovidov** [comment]
Maybe you will be also interested in the internals of `pixz` tool.

**Comment by fibersel** [comment]
> Maybe you will be also interested in the internals of `pixz` tool.

I will keep it in mind, but now I am going to deal with main task.

**Comment by akuzm** [comment]
This PR broke the Fast Test.

https://github.com/ClickHouse/ClickHouse/pull/16908

I'll merge the fix for now, so that we have at least some tests, but it's not entirely correct. CMake should be able to finish when the `xz` submodule is not cloned -- it's not an essential one.

File: .gitmodules
Comment (azat) [inline .gitmodules:197]: Uses spaces for indent, while all other lines uses tabs (since `git submodule add` uses them)

**Pull request merged**
